### ISSUE TITLE
Type inference causing context-bound implicit value not being passed

### ISSUE THREAD

**Issue by jchyb** [lampepfl/dotty#14859]
## Compiler version
3.2.0-RC1-bin-20220308-29073f1-NIGHTLY
also reproducible on 3.0 and 3.1

## Minimized code

```Scala
//> using scala "3.nightly"

trait Tuple[T]
object Tuple {
  def forTuple1[A]: Tuple[Tuple1[A]] = null
}

trait Path

sealed abstract class Matching[+L: Tuple] // if we remove covariance it compiles
case class Matched[L: Tuple](values: L) extends Matching[L]

case class Directive[L]()(implicit val ev: Tuple[L]) {
  def &(other: Directive[L]): Directive[L] = this
  def tflatMap[R: Tuple](f: L => Directive[R]): Directive[R] = new Directive[R]()
}

object Main {
  def tprovide[L: Tuple](values: L): Directive[L] = Directive()

  def pathSuffix[L](p: Path)(implicit tuple: Tuple[L]): Directive[L] = {
    // if we change Matching[L] to Matched[L] it compiles 
    Directive[Tuple1[Matching[L]]]()(Tuple.forTuple1).tflatMap {
      // without "& tprovide(values)" it compiles 
      case Tuple1(Matched(values)) => tprovide(values) & tprovide(values) // the problematic part
    }
  }
}
 
```

## Output

```
[error] ./Bug.scala:27:55: no implicit argument of type Tuple[L$1] was found for an implicit parameter of method tprovide in object Main
[error] 
[error] where:    L$1 is a type in an anonymous function in method pathSuffix with bounds <: L
[error] .
[error] I found:
[error] 
[error]     tuple
[error] 
[error] But parameter tuple does not match type Tuple[L$1].
[error]       case Tuple1(Matched(values)) => tprovide(values) & tprovide(values) // the problematic part
[error]                                                       ^
```

## Expectation
The code should compile - it does compile in Scala 2.13.8 and very minor (commented in the code) changes do allow it to compile it in Scala 3, with the most interesting being the use of a `&` method. Passing the implicit explicitly, like `tprovide(values)(tuple)` does fix it of course. Originally taken and heavily minimized from the akka-http project, during migration.

**Comment by smarter** [comment]
It's possible we could tweak the interaction between type inference and implicit search to find `tuple` here but:
```scala
sealed abstract class Matching[+L: Tuple] // if we remove covariance it compiles
case class Matched[L: Tuple](values: L) extends Matching[L]
```
In general it's a good idea to align variance between parent and children classes: if possible, either make L covariant in both or invariant in both, it gives you more precise types in pattern matching. Though in this case, maybe the compiler should be able to realize that typing `values` as `L$1` where `L$1` is an unknown subtype of `L` is useless and we could just type it as `L`.

**Comment by jchyb** [comment]
Thank you for explaining. I will have to go back to the original code and see if the covariance inconsistency was intentional there. An interesting part is also that if we remove the `&` method, leaving only `case ... => tprovide(values)`, it does actually decide to type `values` as `L` and compiles successfully.

**Comment by smarter** [comment]
Yeah when we typecheck a method call, we first look at the expected type to propagate constraints, so here without the `&`, the compiler sees that `tprovide(values)` must return a value of type `Directive[L]` for the method to typecheck, and therefore this means the type parameter of `tprovide` must be `L` so implicit search succeeds. With the `&`, all we know is that the result of `tprovide(values)` must be some value with an `&` method (or something that can be converted to such a value), so no information from the result type is propagated.

**Comment by smarter** [comment]
(For more details on how type inference works: https://www.youtube.com/watch?v=lMvOykNQ4zs)
